lib/cli/src/commands/run/wasi.rs

```diff
@@ -225,21 +249,161 @@ impl Wasi {
+                    rewind_state.rewind_stack,
+                    rewind_state.store_data,
+                );
+                if res != Errno::Success {
+                    tx.send(Ok(res as i32)).ok();
+                    return;
+                }
+            } else {
+                if let Err(exit_code) = rewind_state
+                    .rewinding_finish::<Memory32>(ctx.env.clone().into_mut(&mut store), trigger_res)
+                {
+                    tx.send(Ok(exit_code.raw())).ok();
+                    return;
+                }
+                let res = rewind::<Memory32>(
+                    ctx.env.clone().into_mut(&mut store),
+                    rewind_state.memory_stack,
+                    rewind_state.rewind_stack,
+                    rewind_state.store_data,
+                );
+                if res != Errno::Success {
+                    tx.send(Ok(res as i32)).ok();
+                    return;
+                }
+            }
+        }
+
+        // Do we want to invoke a function?
+        if let Some(ref invoke) = run.invoke {
+            let res = RunWithPathBuf::inner_module_invoke_function(
+                &mut store,
+                run.instance,
+                run.path.as_path(),
+                invoke,
+                &run.args,
+            )
+            .map(|()| 0);
+
+            ctx.cleanup(&mut store, None);
+
+            tx.send(res).unwrap();
+        } else {
+            let start: Function =
+                RunWithPathBuf::try_find_function(&run.instance, run.path.as_path(), "_start", &[])
+                    .unwrap();
+
+            let result = start.call(&mut store, &[]);
+            Self::handle_result(
+                RunProperties {
+                    ctx,
+                    instance: run.instance,
+                    path: run.path,
+                    invoke: run.invoke,
+                    args: run.args,
+                },
+                store,
+                result,
+                tx,
+            )
+        }
+    }
+
     /// Helper function for handling the result of a Wasi _start function.
-    pub fn handle_result(&self, result: Result<Box<[Value]>, RuntimeError>) -> Result<i32> {
-        match result {
+    pub fn handle_result(
+        run: RunProperties,
+        mut store: Store,
+        result: Result<Box<[Value]>, RuntimeError>,
+        tx: Sender<Result<i32>>,
+    ) {
+        let ctx = run.ctx;
+        let ret: Result<i32> = match result {
             Ok(_) => Ok(0),
             Err(err) => {
-                let err: anyhow::Error = match err.downcast::<WasiError>() {
-                    Ok(WasiError::Exit(exit_code)) => {
-                        return Ok(exit_code.raw());
+                match err.downcast::<WasiError>() {
+                    Ok(WasiError::Exit(exit_code)) => Ok(exit_code.raw()),
+                    Ok(WasiError::DeepSleep(deep)) => {
+                        let pid = ctx.data(&store).pid();
```

Comment:
Yeah, reading this, as mentioned above, we really shouldn't do this for the main thread...

Reply:
I've reverted all this code back to synchronous only, given its isolated to the CLI only anyway its not that big a deal to revert it.